player.cpp

```diff
@@ -17,22 +17,22 @@ static constexpr glm::vec3 PLAYER_POSITION(-550.0f, 0.0f, 0.0f);
 void Player::OnStart()
 {
     mMesh = std::make_unique<Rectangle>(PLAYER_WIDTH, PLAYER_HEIGHT);
-    mPosition = PLAYER_POSITION;
+    mColliderBox = std::make_unique<ColliderBox>(PLAYER_WIDTH, PLAYER_HEIGHT);
+    SetPosition(PLAYER_POSITION);
     mSpeed = PLAYER_SPEED;
+    mName = "Player";
 }
 
 void Player::OnUpdate()
 {
     if (Input::IsKeyPressed(GLFW_KEY_W))
     {
-        mPosition.y += mSpeed;
+        SetPosition(GetPosition() + glm::vec3(0.0f, mSpeed, 0.0f));
     }
     else if (Input::IsKeyPressed(GLFW_KEY_S))
     {
-        mPosition.y -= mSpeed;
+        SetPosition(GetPosition() + glm::vec3(0.0f, -mSpeed, 0.0f));
     }
-
-    std::cout << "Position x: " << mPosition.x << ", y: " << mPosition.y << std::endl;
```

Comment:
An update to your previous comment about printing in the game loop @cjappl . Something will definitely be needed to make this safer. Testing some of the collisions stuff, I was printing out a lot of debugging information. At least 10-15 different print statements per game loop. This caused the FPS to drop to ~75 frames. Curious, I removed all print statements and made the FPS cap unlimited. I measured an average FPS of ~4200. Even this single position print statement here in the game loop will drop the FPS from ~4200 to ~1700, haha. Perhaps a real-time logger is needed???? 😱

Reply:
If only you knew how to implement one?  ;) 

https://www.youtube.com/watch?v=4KFFMGTQIFM

https://github.com/cjappl/rtlog-cpp

I am very curious if all game rendering loops are necessary to be realtime safe. It certainly has all the smells of needing to be realtime safe. You have a hard deadline that if you don't hit it you have consequences (frame drops). I wonder how much game programmers obsess about this stuff like the audio psychos. I may have to investigate a bit, maybe I have something to learn.

Reply:
I imagine it is kind of somewhere in the middle. It would make sense that the game engine would be as realtime safe as possible. But at the same time, for example, Unity exposes the Instantiate() call to create a new GameObject on the fly. This certainly has to allocate somewhere and I had trouble before with this call specifically dropping the FPS because it can be expensive. To fix it, I had to use an object pool to instantiate a bunch of objects at the start, and reuse them later on. So I guess maybe it is more relaxed until it needs to be optimized better? And it seems like Instantiate() is pretty necessary across Unity too. So I don't think there is anyway around it, at least if an object pool is not overkill.

Reply:
Nice. Yeah object pool sounds like a killer solution to this